src/App.tsx

```diff
@@ -1,204 +1,443 @@
+  USER_ID,
+} from './api/todos';
+import { Todo } from './types/Todo';
+import { TodoComponent } from './components/TodoComponent';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [todoTitle, setTodoTitle] = useState('');
+  const [filter, setFilter] = useState<'all' | 'active' | 'completed'>('all');
+  const [titleErrorMessage, setTitleErrorMessage] = useState('');
+  const [addErrorMessage, setAddErrorMessage] = useState('');
+  const [deleteErrorMessage, setDeleteErrorMessage] = useState('');
+  const [updateErrorMessage, setUpdateErrorMessage] = useState('');
+  const [loadingErrorMessage, setLoadingErrorMessage] = useState('');
+
+  const titleField = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    getTodos()
+      .then(fetchedTodos => {
+        setTodos(fetchedTodos);
+      })
+      .catch(error => {
+        setLoadingErrorMessage('Unable to load todos');
+        throw error;
+      });
+
+    if (titleField.current) {
+      titleField.current.focus();
+    }
+  }, []);
+
+  useEffect(() => {
+    let timer: NodeJS.Timeout;
+
+    if (
+      titleErrorMessage ||
+      addErrorMessage ||
+      updateErrorMessage ||
+      loadingErrorMessage
+    ) {
+      timer = setTimeout(() => {
+        setTitleErrorMessage('');
+        setAddErrorMessage('');
+        setUpdateErrorMessage('');
+        setLoadingErrorMessage('');
+      }, 3000);
+    }
+
+    return () => clearTimeout(timer);
+  }, [
+    titleErrorMessage,
+    addErrorMessage,
+    updateErrorMessage,
+    loadingErrorMessage,
+  ]);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  function addTodo({
+    title,
+    completed,
+  }: {
+    title: string;
+    completed: boolean;
+  }) {
+    setTodos(currentTodos => [
+      ...currentTodos,
+      {
+        userId: USER_ID,
+        title,
+        completed,
+        loading: true,
+        id: Date.now() + Math.random(),
+      },
+    ]);
+    createTodos({ userId: USER_ID, title, completed })
+      .then(newTodo => {
+        setTodos(currentTodos =>
+          currentTodos.map((todo, index) => {
+            if (currentTodos.length - 1 === index) {
+              return newTodo;
+            }
+
+            return todo;
+          }),
+        );
+        setTodoTitle('');
+      })
+      .catch(error => {
+        setAddErrorMessage('Unable to add a todo');
+        setTodos(currentTodos => {
+          const currentTodosCopy = [...currentTodos];
+
+          currentTodosCopy.pop();
+
+          return currentTodosCopy;
+        });
+        throw error;
+      });
+  }
+
+  function removeTodo(todoId: number) {
+    deleteTodos(todoId)
+      .then(() => {
+        setTodos(currentTodos =>
+          currentTodos.filter(todo => todo.id !== todoId),
+        );
+      })
+      .catch(error => {
+        setAddErrorMessage('Unable to delete a todo');
+        throw error;
+      });
+  }
+
+  function updateTodo(todoId: number) {
```

Comment:
It is important to stick to one way of creating functions, fix it everywhere
```suggestion
  const updateTodo = (todoId: number) => {
```